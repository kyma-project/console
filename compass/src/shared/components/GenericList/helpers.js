export const filterEntries = (entries, query) => {
  // TEMP
  if (
    entries[0] &&
    (entries[0].hasOwnProperty('name') ||
      entries[0].hasOwnProperty('description'))
  ) {
    return entries.filter(entry => {
      return (
        (entry.name && entry.name.indexOf(query) > -1) ||
        (entry.description && entry.description.indexOf(query) > -1)
      );
    });
  } else {
    return entries;
  }
};
