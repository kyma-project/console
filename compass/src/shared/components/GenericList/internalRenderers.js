import React from 'react';
import { Popover } from 'fundamental-react/Popover';
import { Menu } from 'fundamental-react/Menu';
import { Button } from '@kyma-project/react-components';

export const renderActionElement = (actions, entry) =>
  !Array.isArray(actions) ? null : (
    <Popover
      body={
        <Menu>
          <Menu.List>
            {actions.map(action => (
              <Menu.Item
                onClick={() => action.handler(entry)}
                key={action.name}
              >
                {action.name}
              </Menu.Item>
            ))}
          </Menu.List>
        </Menu>
      }
      control={<Button glyph="vertical-grip" option="light" />}
      placement="bottom-end"
    />
  );
