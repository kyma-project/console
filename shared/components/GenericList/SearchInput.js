import React from 'react';
import PropTypes from 'prop-types';
import 'core-js/es/array/flat-map';

SearchInput.propTypes = {
  searchQuery: PropTypes.string,
  entries: PropTypes.arrayOf(PropTypes.object.isRequired),
  handleQueryChange: PropTypes.func.isRequired,
  suggestionProperties: PropTypes.arrayOf(PropTypes.string.isRequired)
    .isRequired,
};

export default function SearchInput({
  searchQuery,
  filteredEntries,
  handleQueryChange,
  suggestionProperties,
}) {
  const [isSearchHidden, setSearchHidden] = React.useState(true);
  const searchInputRef = React.useRef();

  const renderSearchList = entries => {
    const suggestions = getSearchSuggestions(entries);

    if (!suggestions.length) {
      return (
        <li
          key="no-entries"
          className="fd-menu__item fd-menu__item--no-entries"
        >
          No entries found
        </li>
      );
    }

    return suggestions.map(suggestion => (
      <li
        onClick={() => handleQueryChange(suggestion)}
        key={suggestion}
        className="fd-menu__item"
      >
        {suggestion}
      </li>
    ));
  };

  const getSearchSuggestions = entries => {
    const suggestions = entries
      .flatMap(entry => {
        return suggestionProperties.map(property => {
          const entryValue = entry[property];
          if (
            entryValue &&
            entryValue
              .toString()
              .toLowerCase()
              .includes(searchQuery.toLowerCase())
          ) {
            return entryValue;
          }
          return null;
        });
      })
      .filter(suggestion => suggestion);
    return Array.from(new Set(suggestions));
  };

  const openSearchList = () => {
    setSearchHidden(false);
    setImmediate(() => {
      const inputField = searchInputRef.current;
      inputField.focus();
    });
  };

  const checkForEscapeKey = e => {
    const ESCAPE_KEY_CODE = 27;
    if (e.keyCode === ESCAPE_KEY_CODE) {
      setSearchHidden(true);
    }
  };

  const showControl = isSearchHidden && !searchQuery;
  return (
    <section className="generic-list-search" role="search">
      <div className="fd-popover">
        <div className="fd-popover__control">
          <div className="fd-combobox-control">
            <input
              aria-label="search-input"
              ref={searchInputRef}
              type="text"
              value={searchQuery}
              onBlur={() => setSearchHidden(true)}
              onFocus={() => setSearchHidden(false)}
              onChange={e => handleQueryChange(e.target.value)}
              onKeyPress={checkForEscapeKey}
              style={{ display: showControl ? 'none' : 'initial' }}
            />
            {!!searchQuery && (
              <div
                className="fd-popover__body fd-popover__body--no-arrow"
                aria-hidden={isSearchHidden}
              >
                <nav className="fd-menu">
                  <ul className="fd-menu__list">
                    {renderSearchList(filteredEntries)}
                  </ul>
                </nav>
              </div>
            )}
          </div>
        </div>
      </div>
      {showControl && (
        <button
          className="fd-button--light sap-icon--search"
          onClick={openSearchList}
          aria-label="open-search"
        />
      )}
    </section>
  );
}
