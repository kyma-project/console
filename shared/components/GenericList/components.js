import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Button } from 'fundamental-react';

import ListActions from '../ListActions/ListActions';

export const BodyFallback = ({ children }) => (
  <tr>
    <td colSpan="100%">
      <div className="body-fallback">{children}</div>
    </td>
  </tr>
);

export const HeaderRenderer = ({ actions, headerRenderer }) => {
  let emptyColumn = [];
  if (actions.length) {
    emptyColumn = [<th key="actions-column" aria-label="actions-column"></th>];
  }
  return [
    headerRenderer().map((h, index) => <th key={h || index}>{h}</th>),
    ...emptyColumn,
  ];
};

export const RowRenderer = ({ entry, actions, rowRenderer, ...others }) => {
  const filteredActions = actions.filter(a =>
    a.skipAction ? !a.skipAction(entry) : true,
  );
  const resolvedRowRenderer = rowRenderer(entry);

  if (Array.isArray(resolvedRowRenderer)) {
    return (
      <DefaultRowRenderer
        {...others}
        entry={entry}
        actions={filteredActions}
        rowRenderer={resolvedRowRenderer}
      />
    );
  }
  return (
    <CollapsedRowRenderer
      {...others}
      entry={entry}
      actions={filteredActions}
      rowRenderer={resolvedRowRenderer}
    />
  );
};

const DefaultRowRenderer = ({
  key,
  entry,
  actions,
  rowRenderer,
  actionsStandaloneItems,
}) => {
  const cells = rowRenderer.map((cell, id) => <td key={id}>{cell}</td>);
  const actionsCell = (
    <td>
      <ListActions
        actions={actions}
        entry={entry}
        standaloneItems={actionsStandaloneItems}
      />
    </td>
  );

  return (
    <tr key={key}>
      {cells}
      {!!actions.length && actionsCell}
    </tr>
  );
};

const CollapsedRowRenderer = ({
  key,
  entry,
  actions,
  rowRenderer: { cells, collapseContent, showCollapseControl = true },
  actionsStandaloneItems,
}) => {
  const [isOpen, setOpen] = useState(false);

  const rowRenderer = [
    showCollapseControl ? (
      <Button
        data-testid={isOpen ? 'collapse-button-open' : 'collapse-button-close'}
        glyph={isOpen ? 'navigation-up-arrow' : 'navigation-down-arrow'}
        option="light"
        onClick={() => setOpen(!isOpen)}
        compact={true}
        typeAttr="button"
      />
    ) : (
      <></>
    ),
    ...cells,
  ];

  const defaultRow = (
    <DefaultRowRenderer
      key={key}
      entry={entry}
      actions={actions}
      rowRenderer={rowRenderer}
      actionsStandaloneItems={actionsStandaloneItems}
    />
  );

  return (
    <>
      {defaultRow}
      {isOpen ? (
        <tr role="row" data-testid="collapse-content">
          {collapseContent}
        </tr>
      ) : null}
    </>
  );
};
