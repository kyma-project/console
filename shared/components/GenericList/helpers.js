const filterEntry = (entry, query, searchProperties) => {
  if (!query) {
    return true;
  }

  if (!Object.keys(searchProperties).length) {
    return false;
  }
  for (const property of searchProperties) {
    if (entry.hasOwnProperty(property)) {
      const value = entry[property];
      // apply to string to include numbers
      if (
        value &&
        value
          .toString()
          .toLowerCase()
          .indexOf(query.toLowerCase()) !== -1
      ) {
        return true;
      }
    }
  }
  return false;
};

export const filterEntries = (entries, query, searchProperties) => {
  return entries.filter(entry => filterEntry(entry, query, searchProperties));
};
