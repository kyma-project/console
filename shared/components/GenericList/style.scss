.generic-list {

    .fd-panel__body {
        padding: 0;
    }

    .fd-panel__header {
        display: flex;
        justify-content: space-between;

        .fd-panel__title {
            font-size: 16px;
        }

        > section > * {
            display: inline-block;
        }
        
    }
}

.generic-list-search {
    position: relative;
    top: 2px;
    margin-right: 8px;

    .fd-menu__list {
        max-height: 50vh;
        overflow-y: scroll;

        .fd-menu__item {
            overflow: hidden;
            text-overflow: ellipsis;
        }

        .fd-menu__item--no-entries {
            font-style: italic;
        }
    }

    .control {
        &--open {
            width: unset;
            transition: 0.4s;
        }

        &--close {
            transition: 0.4s;
            width: 0;
            margin: 0;
            padding: 0;
            border: none;
        }
    }
}
