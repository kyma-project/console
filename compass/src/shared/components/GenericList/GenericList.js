import React from 'react';
import PropTypes from 'prop-types';

import { Panel } from 'fundamental-react/lib/Panel';

import {
  Search,
  TableWithActionsToolbar,
  TableWithActionsList,
} from '@kyma-project/react-components';

import { filterEntries } from './helpers';
import { renderActionElement } from './internalRenderers';

class GenericList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      entries: props.entries,
      filteredEntries: props.entries,
    };
  }

  headerRenderer = entries => {
    if (this.props.actions) {
      return [...this.props.headerRenderer(entries), ''];
    } else {
      return this.props.headerRenderer(entries);
    }
  };

  rowRenderer = entry => {
    if (this.props.actions) {
      return [
        ...this.props.rowRenderer(entry),
        renderActionElement(this.props.actions, entry),
      ];
    } else {
      return this.props.rowRenderer(entry);
    }
  };

  renderSearchList = entries => {
    const filteredEntries =
      entries && entries.length
        ? entries.map(entry => {
            return entry.name !== undefined && entry.description !== undefined
              ? [
                  {
                    text: entry.name.substring(0, 18),
                    callback: () => this.handleQueryChange(entry.name),
                  },
                  {
                    text: entry.description.substring(0, 18),
                    callback: () => this.handleQueryChange(entry.description),
                  },
                ]
              : entries;
          })
        : null;
    return Array.isArray(filteredEntries) ? filteredEntries.flat() : entries;
  };

  handleQueryChange = event => {
    const searchTerm = event.target.value;
    this.setState(prevState => ({
      filteredEntries: filterEntries(prevState.entries, searchTerm),
    }));
  };

  render() {
    const { filteredEntries } = this.state;
    const { extraHeaderContent } = this.props;

    const headerActions = filteredEntries => (
      <>
        {extraHeaderContent}
        {/* {this.processFilterElement(allFilters)} */}
        <Search
          placeholder="Search..."
          onChange={this.handleQueryChange}
          searchList={this.renderSearchList(filteredEntries)}
        />
      </>
    );

    return (
      <Panel className="fd-panel--no-background">
        <TableWithActionsToolbar
          title={this.props.title}
          description={this.props.description}
          children={headerActions(filteredEntries)}
        />

        <Panel.Body>
          <TableWithActionsList
            entries={filteredEntries}
            headerRenderer={this.headerRenderer}
            rowRenderer={this.rowRenderer}
          />
        </Panel.Body>
      </Panel>
    );
  }
}
export default GenericList;

GenericList.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  entries: PropTypes.arrayOf(PropTypes.object),
  headerRenderer: PropTypes.func.isRequired,
  rowRenderer: PropTypes.func.isRequired,
  actions: PropTypes.arrayOf(
    PropTypes.shape({ name: PropTypes.string, handler: PropTypes.func }),
  ),
  extraHeaderContent: PropTypes.node,
};
