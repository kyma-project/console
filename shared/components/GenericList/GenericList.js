import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import SearchInput from './SearchInput';
import { Panel } from 'fundamental-react/Panel';
import { filterEntries } from './helpers';
import ListActions from '../ListActions/ListActions';
import './GenericList.scss';

const NotFoundMessage = ({ children }) => (
  <td colSpan="100%">
    <p className="not-found-message">{children}</p>
  </td>
);

const HeaderRenderer = ({ entries, actions, headerRenderer }) => {
  let emptyColumn = [];
  if (actions.length) {
    emptyColumn = [<th key="actions-column" aria-label="actions-column"></th>];
  }
  return [headerRenderer().map(h => <th key={h}>{h}</th>), ...emptyColumn];
};

const RowRenderer = ({
  entry,
  actions,
  rowRenderer,
  actionsStandaloneItems,
}) => {
  const filteredActions = actions.filter(a =>
    a.skipAction ? !a.skipAction(entry) : true,
  );

  const cells = rowRenderer(entry).map((cell, id) => <td key={id}>{cell}</td>);
  const actionsCell = (
    <td>
      <ListActions
        actions={filteredActions}
        entry={entry}
        standaloneItems={actionsStandaloneItems}
      />
    </td>
  );

  return (
    <>
      {cells}
      {!!filteredActions.length && actionsCell}
    </>
  );
};

export const GenericList = ({
  entries,
  actions,
  title,
  headerRenderer,
  rowRenderer,
  notFoundMessage,
  extraHeaderContent,
  showSearchField,
  textSearchProperties,
  actionsStandaloneItems,
}) => {
  const [filteredEntries, setFilteredEntries] = useState(entries);
  const [searchQuery, setSearchQuery] = useState('');

  useEffect(() => {
    setFilteredEntries(
      filterEntries([...entries], searchQuery, textSearchProperties),
    );
  }, [searchQuery, setFilteredEntries, entries]);

  const headerActions = (
    <section className="generic-list__search">
      {showSearchField && (
        <SearchInput
          searchQuery={searchQuery}
          filteredEntries={filteredEntries}
          handleQueryChange={setSearchQuery}
          suggestionProperties={textSearchProperties}
        />
      )}
      {extraHeaderContent}
    </section>
  );

  return (
    <Panel className="fd-has-margin-m generic-list">
      <Panel.Header className="fd-has-padding-xs">
        <Panel.Head title={title} />
        <Panel.Actions>{headerActions}</Panel.Actions>
      </Panel.Header>

      <Panel.Body>
        <table className="fd-table">
          <thead>
            <tr>
              <HeaderRenderer
                entries={entries}
                actions={actions}
                headerRenderer={headerRenderer}
              />
            </tr>
          </thead>
          <tbody>
            {filteredEntries.length ? (
              filteredEntries.map((e, index) => (
                <tr role="row" key={e.id || e.name || index}>
                  <RowRenderer
                    entry={e}
                    actions={actions}
                    actionsStandaloneItems={actionsStandaloneItems}
                    rowRenderer={rowRenderer}
                  />
                </tr>
              ))
            ) : (
              <tr>
                <NotFoundMessage>{notFoundMessage}</NotFoundMessage>
              </tr>
            )}
          </tbody>
        </table>
      </Panel.Body>
    </Panel>
  );
};

GenericList.Actions = ListActions;

GenericList.propTypes = {
  title: PropTypes.string,
  entries: PropTypes.arrayOf(PropTypes.object).isRequired,
  headerRenderer: PropTypes.func.isRequired,
  rowRenderer: PropTypes.func.isRequired,
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      handler: PropTypes.func.isRequired,
      skipAction: PropTypes.func,
    }),
  ).isRequired,
  extraHeaderContent: PropTypes.node,
  showSearchField: PropTypes.bool,
  notFoundMessage: PropTypes.string,
  textSearchProperties: PropTypes.arrayOf(PropTypes.string.isRequired),
  actionsStandaloneItems: PropTypes.number,
};

GenericList.defaultProps = {
  notFoundMessage: 'No entries found',
  actions: [],
  showSearchField: true,
  textSearchProperties: ['name', 'description'],
};
