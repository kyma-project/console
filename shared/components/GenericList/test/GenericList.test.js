import React from 'react';
import { GenericList } from '../GenericList';

import 'core-js/es/array/flat-map';
import { render, fireEvent } from '@testing-library/react';

describe('GenericList', () => {
  const defaultNotFoundText = 'No entries found';

  const mockHeaderRenderer = entries => ['Id', 'Name', 'description'];
  const mockEntryRenderer = entry => [entry.id, entry.name, entry.description];

  const mockEntries = [
    { id: '1', name: 'first_entry', description: 'testdescription1' },
    { id: '2', name: 'second_entry', description: 'testdescription2' },
    { id: '3', name: 'THIRD_ENTRY', description: 'testdescription3' },
  ];

  it('Renders with minimal props', async () => {
    const { getByText } = render(
      <GenericList
        title=""
        entries={[]}
        headerRenderer={() => []}
        rowRenderer={() => []}
      />,
    );
    await getByText(defaultNotFoundText);
  });

  it('Renders custom notFoundMessage props', async () => {
    const notFoundMessage = 'abcd';
    const { getByText } = render(
      <GenericList
        title=""
        entries={[]}
        headerRenderer={() => []}
        rowRenderer={() => []}
        notFoundMessage={notFoundMessage}
      />,
    );
    await getByText(notFoundMessage);
  });

  it('Renders title', async () => {
    const title = 'title';
    const { getByText } = render(
      <GenericList
        title={title}
        entries={[]}
        headerRenderer={() => []}
        rowRenderer={() => []}
      />,
    );
    await getByText(title);
  });

  describe('Actions', () => {
    it("Renders actions button when 'actions' prop is provided", () => {
      const actions = [{ name: 'testaction', handler: () => {} }];
      const { getAllByLabelText } = render(
        <GenericList
          headerRenderer={() => []}
          rowRenderer={() => []}
          actions={actions}
          entries={mockEntries}
        />,
      );
      const actionButtons = getAllByLabelText(actions[0].name);
      expect(actionButtons.length).toBe(mockEntries.length);
    });

    it("Skips rendering actions when 'actions' prop passes skipAction() call", () => {
      const actions = [
        { name: 'skip it', handler: () => {}, skipAction: () => true },
        {
          name: 'no skipping here',
          handler: () => {},
          skipAction: () => false,
        },
      ];
      const { queryByLabelText } = render(
        <GenericList
          headerRenderer={() => []}
          rowRenderer={() => []}
          actions={actions}
          entries={[{ id: '23' }]}
        />,
      );
      expect(queryByLabelText(actions[0].name)).toBeNull();
      expect(queryByLabelText(actions[1].name)).toBeTruthy();
    });

    it('Renders extra column in header when only actions are set', () => {
      const actions = [{ name: 'testaction', handler: () => {} }];
      const { queryByLabelText, rerender } = render(
        <GenericList
          headerRenderer={() => []}
          rowRenderer={() => []}
          actions={actions}
          entries={mockEntries}
        />,
      );

      expect(queryByLabelText('actions-column')).toBeInTheDocument();

      rerender(
        <GenericList
          headerRenderer={() => []}
          rowRenderer={() => []}
          entries={mockEntries}
        />,
      );

      expect(queryByLabelText('actions-column')).not.toBeInTheDocument();
    });
  });

  it('Renders entries', async () => {
    const { getByText } = render(
      <GenericList
        entries={mockEntries}
        headerRenderer={mockHeaderRenderer}
        rowRenderer={mockEntryRenderer}
      />,
    );

    mockEntries.forEach(entry =>
      Object.keys(entry).forEach(key => getByText(entry[key])),
    );
  });

  it('Renders custom data using custom entryRenderer', async () => {
    const customEntryRenderer = entry => [entry.name, 'maskopatol'];

    const { queryByText } = render(
      <GenericList
        entries={[mockEntries[0]]}
        headerRenderer={mockHeaderRenderer}
        rowRenderer={customEntryRenderer}
      />,
    );

    expect(await queryByText(mockEntries[0].name)).toBeTruthy();
  });

  it('Renders headers', async () => {
    const { getByText } = render(
      <GenericList
        entries={mockEntries}
        headerRenderer={mockHeaderRenderer}
        rowRenderer={mockEntryRenderer}
      />,
    );

    mockHeaderRenderer().forEach(async header => await getByText(header));
  });

  it('Renders extreaHeaderContent', async () => {
    const content = 'wow this is so extra!';
    const { getByText } = render(
      <GenericList
        entries={mockEntries}
        headerRenderer={mockHeaderRenderer}
        rowRenderer={mockEntryRenderer}
        extraHeaderContent={<span>{content}</span>}
      />,
    );

    await getByText(content);
  });

  describe('Search', () => {
    it('Show search field by default', async () => {
      const { getByRole } = render(
        <GenericList
          entries={mockEntries}
          headerRenderer={mockHeaderRenderer}
          rowRenderer={mockEntryRenderer}
        />,
      );

      await getByRole('search');
    });

    it("Desn't show search field when showSearchField is set to false", async () => {
      const { queryByRole } = render(
        <GenericList
          entries={mockEntries}
          headerRenderer={mockHeaderRenderer}
          rowRenderer={mockEntryRenderer}
          showSearchField={false}
        />,
      );

      expect(await queryByRole('search')).toBeNull();
    });

    it('Finds proper entries when search text is entered', async () => {
      const searchText = 'first';

      const { queryAllByRole, getByLabelText } = render(
        <GenericList
          entries={mockEntries}
          headerRenderer={mockHeaderRenderer}
          rowRenderer={mockEntryRenderer}
        />,
      );

      expect(await queryAllByRole('row')).toHaveLength(mockEntries.length + 1); // header + {mockEntries.length} rows

      const searchInput = await getByLabelText('search-input');
      fireEvent.change(searchInput, { target: { value: searchText } });

      expect(await queryAllByRole('row')).toHaveLength(2); // header + one row
    });

    it('Search is case insensitive', async () => {
      let searchText = 'third';

      const { queryAllByRole, getByLabelText } = render(
        <GenericList
          entries={mockEntries}
          headerRenderer={mockHeaderRenderer}
          rowRenderer={mockEntryRenderer}
        />,
      );

      expect(await queryAllByRole('row')).toHaveLength(mockEntries.length + 1); // header + {mockEntries.length} rows

      let searchInput = await getByLabelText('search-input');
      fireEvent.change(searchInput, { target: { value: searchText } });

      expect(await queryAllByRole('row')).toHaveLength(2); // header + one row

      searchText = 'THIRD';
      searchInput = await getByLabelText('search-input');
      fireEvent.change(searchInput, { target: { value: searchText } });

      expect(await queryAllByRole('row')).toHaveLength(2); // header + one row
    });

    it('Shows no entries message when there are no results', async () => {
      const searchText = "Do you really can't find it?";
      const notFoundMessage = 'Yes, sorry';

      const { queryAllByRole, getByLabelText, getByText } = render(
        <GenericList
          entries={mockEntries}
          headerRenderer={mockHeaderRenderer}
          rowRenderer={mockEntryRenderer}
          notFoundMessage={notFoundMessage} // because the default text is the same as the one displayed by Search suggestions
        />,
      );

      expect(await queryAllByRole('row')).toHaveLength(mockEntries.length + 1); // header + {mockEntries.length} rows

      const searchInput = await getByLabelText('search-input');
      fireEvent.change(searchInput, { target: { value: searchText } });

      expect(await queryAllByRole('row')).toHaveLength(2); // header + NotFoundMessage dedicated row
      await getByText(notFoundMessage);
    });
  });
});
