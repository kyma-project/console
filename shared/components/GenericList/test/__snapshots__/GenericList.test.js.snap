// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericList Renders actions button when 'actions' prop is provided 1`] = `
<div
  className="fd-panel fd-has-margin-m generic-list"
>
  <div
    className="fd-panel__header fd-has-padding-xs"
  >
    <div
      className="fd-panel__head"
    />
    <section
      className="generic-list__search"
    >
      <section
        className="generic-list-search"
      >
        <div
          className="fd-popover"
        >
          <div
            className="fd-popover__control"
          >
            <div
              className="fd-combobox-control"
            >
              <input
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyPress={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <button
          className="fd-button--light sap-icon--search"
          onClick={[Function]}
        />
      </section>
    </section>
  </div>
  <div
    className="fd-panel__body"
  >
    <div
      className="fd-panel sc-caSCKo jjNSwx"
    >
      <div
        className="fd-panel__body sc-TOsTZ iphcEB"
      >
        <table
          className="fd-table sc-kgAjT EPChu"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                first_entry
              </td>
              <td>
                <div
                  className="fd-popover"
                >
                  <div
                    onFocus={[Function]}
                    onMouseDown={[Function]}
                    onTouchStart={[Function]}
                  >
                    <div
                      className="fd-popover__control"
                    >
                      <button
                        className="fd-button--light sap-icon--vertical-grip"
                        onClick={[Function]}
                      />
                    </div>
                  </div>
                </div>
              </td>
            </tr>
            <tr>
              <td>
                second_entry
              </td>
              <td>
                <div
                  className="fd-popover"
                >
                  <div
                    onFocus={[Function]}
                    onMouseDown={[Function]}
                    onTouchStart={[Function]}
                  >
                    <div
                      className="fd-popover__control"
                    >
                      <button
                        className="fd-button--light sap-icon--vertical-grip"
                        onClick={[Function]}
                      />
                    </div>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`GenericList Renders title and description 1`] = `
<div
  className="fd-panel fd-has-margin-m generic-list"
>
  <div
    className="fd-panel__header fd-has-padding-xs"
  >
    <div
      className="fd-panel__head"
    >
      <h3
        className="fd-panel__title"
      >
        title
      </h3>
    </div>
    <section
      className="generic-list__search"
    >
      <section
        className="generic-list-search"
      >
        <div
          className="fd-popover"
        >
          <div
            className="fd-popover__control"
          >
            <div
              className="fd-combobox-control"
            >
              <input
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyPress={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <button
          className="fd-button--light sap-icon--search"
          onClick={[Function]}
        />
      </section>
    </section>
  </div>
  <div
    className="fd-panel__body"
  >
    <div
      className="fd-panel sc-caSCKo jjNSwx"
    >
      <div
        className="fd-panel__body sc-TOsTZ iphcEB"
      >
        <table
          className="fd-table sc-kgAjT EPChu"
        >
          <thead>
            <tr />
          </thead>
          <tbody />
        </table>
        <p
          className="sc-cJSrbW gXxzRq"
        >
          There are no items to show
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`GenericList Renders with minimal props 1`] = `
<div
  className="fd-panel fd-has-margin-m generic-list"
>
  <div
    className="fd-panel__header fd-has-padding-xs"
  >
    <div
      className="fd-panel__head"
    />
    <section
      className="generic-list__search"
    >
      <section
        className="generic-list-search"
      >
        <div
          className="fd-popover"
        >
          <div
            className="fd-popover__control"
          >
            <div
              className="fd-combobox-control"
            >
              <input
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyPress={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <button
          className="fd-button--light sap-icon--search"
          onClick={[Function]}
        />
      </section>
    </section>
  </div>
  <div
    className="fd-panel__body"
  >
    <div
      className="fd-panel sc-caSCKo jjNSwx"
    >
      <div
        className="fd-panel__body sc-TOsTZ iphcEB"
      >
        <table
          className="fd-table sc-kgAjT EPChu"
        >
          <thead>
            <tr />
          </thead>
          <tbody />
        </table>
        <p
          className="sc-cJSrbW gXxzRq"
        >
          There are no items to show
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`GenericList Skips rendering actions when 'actions' prop passes skipAction() call 1`] = `
<div
  className="fd-panel fd-has-margin-m generic-list"
>
  <div
    className="fd-panel__header fd-has-padding-xs"
  >
    <div
      className="fd-panel__head"
    />
    <section
      className="generic-list__search"
    >
      <section
        className="generic-list-search"
      >
        <div
          className="fd-popover"
        >
          <div
            className="fd-popover__control"
          >
            <div
              className="fd-combobox-control"
            >
              <input
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyPress={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <button
          className="fd-button--light sap-icon--search"
          onClick={[Function]}
        />
      </section>
    </section>
  </div>
  <div
    className="fd-panel__body"
  >
    <div
      className="fd-panel sc-caSCKo jjNSwx"
    >
      <div
        className="fd-panel__body sc-TOsTZ iphcEB"
      >
        <table
          className="fd-table sc-kgAjT EPChu"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                first_entry
              </td>
            </tr>
            <tr>
              <td>
                second_entry
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
